### PR TITLE
Add General_Category and further predicates

### LINKED ISSUE #38: Add support for General_Category property
This property is useful to implement predicates as in `Data.Char`.

### LINKED ISSUE #39: Code point ranges are not parsed in UnicodeData.txt
See: https://www.unicode.org/reports/tr44/#Code_Point_Ranges

> For backward compatibility, ranges in the file UnicodeData.txt are specified by entries for the start and end characters of the range, rather than by the form "X..Y". The start character is indicated by a range identifier, followed by a comma and the string "First", in angle brackets. This entry takes the place of a regular character name in field 1 for that line. The end character is indicated on the next line with the same range identifier, followed by a comma and the string "Last", in angle brackets:
>
>    4E00;<CJK Ideograph, First>;Lo;0;L;;;;;N;;;;;
>    9FEF;<CJK Ideograph, Last>;Lo;0;L;;;;;N;;;;;
>  
> For character ranges using this convention, the names of all characters in the range are algorithmically derivable. See Section 4.8, Name in [Unicode] for more information on derivation of character names for such ranges.

### PR DESCRIPTION
**Pull request by wismill**
- Parser
    - Fix parsing of ranges in `UnicodeData.txt`
    - Generate `lib/Unicode/Internal/Char/UnicodeData/GeneralCategory.hs`
- Library
    - Add `GeneralCategory` data type and corresponding `generalCategoryAbbr, generalCategory` functions.
    - Add the following functions: `isAlpha, isAlphaNum, isControl, isMark, isNumber, isPrint, isPunctuation, isSeparator, isSymbol`.
    - **Breaking change:** `isLetter` is renamed `isAlphabetic` and `isSpace` is renamed `isWhiteSpace`.
    - **Breaking change:** `isLetter` and `isSpace` now match the base’s Data.Char behaviour.
    - Add `lookupIntN` in `Unicode.Internal.Bits` for low-level stuff.
    - Re-export some functions from `Data.Char` in order to make `Unicode.Char` a drop-in replacement.
- Test
   - Add tests for `Unicode.Char` to ensure we get the same result than `Data.Char`.
     Note that this true only if the compiler used have have the same Unicode version as this package.

Fixes #38, Fixes #39

**Note:** `ucd.sh generate` should be re-run if PR #36 is merge (before or after this PR).

### CODE REVIEW

**Comment by harendra-kumar** [comment]
@Bodigrim would you like to review the changes to `lib/Unicode/Internal/Bits.hs`?

**Comment by Bodigrim** [comment]
@wismill I appreciate the effort, but I think sub-byte operations are not worth it. There are 31 constructors, so you need 5 bits per one value. It's most likely faster just use a byte per value, but fetch all data in one go instead of conditional fetching of one or two words.

**Comment by wismill** [comment]
@Bodigrim Ok, I will check that. Meanwhile, I added a benchmark and disable tests with incompatible GHC versions.

**Comment by wismill** [comment]
@Bodigrim I have implemented your suggestion. It is roughly 60 to 80% faster than my previous implementation!

If my benchmark is correct, we have a nice speedup of 16x compared to some functions from `Data.Char`.

```
All
  Unicode.Char.Case
    isLower
      base:           OK (0.52s)
         17 ms ± 585 μs
      unicode-data:   OK (0.33s)
        315 μs ±  14 μs
    isUpper
      base:           OK (2.07s)
         16 ms ± 879 μs
      unicode-data:   OK (0.31s)
        314 μs ±  12 μs
  Unicode.Char.General
    generalCategory
      base:           OK (0.86s)
        122 ms ± 4.2 ms
      unicode-data:   OK (0.33s)
        109 ms ± 3.4 ms
    isAlpa
      base:           OK (0.24s)
         16 ms ± 795 μs
      unicode-data:   OK (0.25s)
        945 μs ±  53 μs
    isAlpabetic
      unicode-data:   OK (0.32s)
        314 μs ±  11 μs
    isAlpaNum
      base:           OK (1.01s)
         16 ms ± 263 μs
      unicode-data:   OK (0.49s)
        946 μs ±  35 μs
    isControl
      base:           OK (0.49s)
         16 ms ± 424 μs
      unicode-data:   OK (0.20s)
        783 μs ±  46 μs
    isLetter
      base:           OK (0.20s)
        783 μs ±  43 μs
      unicode-data:   OK (0.20s)
        784 μs ±  44 μs
    isMark
      base:           OK (0.25s)
         16 ms ± 693 μs
      unicode-data:   OK (0.20s)
        783 μs ±  44 μs
    isNumber
      base:           OK (0.52s)
         17 ms ± 843 μs
      unicode-data:   OK (0.49s)
        947 μs ±  28 μs
    isPrint
      base:           OK (0.50s)
         16 ms ± 901 μs
      unicode-data:   OK (0.24s)
        943 μs ±  53 μs
    isPunctuation
      base:           OK (0.47s)
         15 ms ± 813 μs
      unicode-data:   OK (0.20s)
        783 μs ±  42 μs
    isSeparator
      base:           OK (0.51s)
         16 ms ± 644 μs
      unicode-data:   OK (0.24s)
        949 μs ±  50 μs
    isSpace
      base:           OK (0.40s)
        1.6 ms ±  43 μs
      unicode-data:   OK (0.24s)
        1.9 ms ± 109 μs
    isSymbol
      base:           OK (0.21s)
         15 ms ± 714 μs
      unicode-data:   OK (0.24s)
        940 μs ±  49 μs
    isWhiteSpace
      unicode-data:   OK (0.32s)
        315 μs ±  12 μs
    isHangul
      unicode-data:   OK (0.32s)
        314 μs ±  11 μs
    isHangulLV
      unicode-data:   OK (0.32s)
        314 μs ±  12 μs
    isJamo
      unicode-data:   OK (0.32s)
        314 μs ±  11 μs
    jamoLIndex
      unicode-data:   OK (0.32s)
        314 μs ±  11 μs
    jamoVIndex
      unicode-data:   OK (0.32s)
        314 μs ±  11 μs
    jamoTIndex
      unicode-data:   OK (0.32s)
        314 μs ±  12 μs
  Unicode.Char.Identifiers
    isIDContinue
      unicode-data:   OK (0.33s)
        314 μs ±  11 μs
    isIDStart
      unicode-data:   OK (0.33s)
        314 μs ±  12 μs
    isXIDContinue
      unicode-data:   OK (0.32s)
        314 μs ±  11 μs
    isXIDStart
      unicode-data:   OK (0.32s)
        629 μs ±  24 μs
    isPatternSyntax
      unicode-data:   OK (0.33s)
        314 μs ±  13 μs
    isPatternWhitespace
      unicode-data:   OK (0.33s)
        315 μs ±  13 μs
  Unicode.Char.Normalization
    isCombining
      unicode-data:   OK (0.33s)
        314 μs ±  11 μs
    combiningClass
      unicode-data:   OK (0.21s)
        3.3 ms ± 188 μs
    isCombiningStarter
      unicode-data:   OK (0.32s)
        314 μs ±  11 μs
    isDecomposable
      Canonical
        unicode-data: OK (0.33s)
          315 μs ±  13 μs
      Kompat
        unicode-data: OK (0.33s)
          315 μs ±  12 μs
    decomposeHangul
      unicode-data:   OK (0.32s)
        314 μs ±  11 μs

All 48 tests passed (18.65s)
```

**Comment by wismill** [comment]
I have merged latest changes from `master`. 

All tests successful, except Ubuntu+GHC7103, which seems to have a linker issue. Could you check it?

I propose to bump the package version from 0.2 to 0.3, as there are some breaking changes.

**Comment by wismill** [comment]
@harendra-kumar @adithyaov @Bodigrim I have merged master, added support for big-endian architectures (based on @Bodigrim work). There are some CI failures, although they seem unrelated to this PR.

File: lib/Unicode/Internal/Bits.hs [hunk 1]
```diff
@@ -49,3 +54,30 @@ lookupBit64 addr# (I# index#) = W# (word## `and#` bitMask##) /= 0
...
+* @ceiling (addr + (n * 8))@ is legally accessible @Word8@.
+-}
+lookupIntN
+  :: Addr# -- ^ Bitmap address
+  -> Int   -- ^ Word index
+  -> Int   -- ^ Resulting word as 'Int'
+lookupIntN addr# (I# index#) = I# (word2Int# word##)
+  where
+#if MIN_VERSION_base(4,16,0)
+#ifdef WORDS_BIGENDIAN
```
Comment (Bodigrim) [inline lib/Unicode/Internal/Bits.hs:72]: I don't think this is correct: there is nothing to swap in a single byte. The situation is different from `lookupBit64`, where we index a whole word and must swap bytes inside. 

Since you created tests, you can add big-endian CI config similar to https://github.com/composewell/unicode-transforms/pull/75 to check that things work well.

Comment (wismill) [reply lib/Unicode/Internal/Bits.hs]: I overlooked this one, sorry. I fixed it by reverting the corresponding commit. 

However I do not know how to implement the CI:
- The tests just compare functions to their equivalent in `Data.Char`.
- In order to work, we need a version of GHC using the same version of Unicode than this package. 
- However, currently there is only GHC 9.2.1, which is not available in Ubuntu 20.04 nor using [hvr’s ppa](https://launchpad.net/~hvr/+archive/ubuntu/ghc).

**Comment by harendra-kumar** [comment]
@adithyaov can you help with appveyor CI failure and GHC 9.2.1 CI, I think you fixed a similar issue elsewhere. GHC 9.2.1 CI example can be found [here](https://github.com/composewell/streamly/blob/fb5cc5401cfa2cdc4f03aa71b82e277d06c4b2d8/.github/workflows/haskell.yml).

File: lib/Unicode/Internal/Bits.hs (hunk 1 above)
Comment (Bodigrim) [reply lib/Unicode/Internal/Bits.hs]: Right, you want be able to test big-endian GHC 9.2.1, because there is simply no bindist at the moment. It should be fine as is.

**Comment by Bodigrim** [comment]
Appveyor failure is a known issue with `lts-18.17`, try to update to `lts-18.18` or newer.

**Review by Bodigrim — APPROVED** [review]

File: lib/Unicode/Internal/Bits.hs (hunk 1 above)
Comment (wismill) [reply lib/Unicode/Internal/Bits.hs]: Should I adapt your CI using same GHC from Ubuntu, meaning tests will be skipped? Or is better to spare CI till we get the ghc and create an issue to keep track ?

Comment (Bodigrim) [reply lib/Unicode/Internal/Bits.hs]: Nevermind, GHC 9.2.1 bindist for big endian is not yet happening in the nearest years.

**Comment by harendra-kumar** [comment]
I will take a better look at the APIs, naming, structuring etc once I get some time.

File: test/Unicode/CharSpec.hs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+#endif
+  describe' "Unicode general categories" do
+    it "generalCategory" do
+      -- Note: we cannot compare the categories directly, so use 'show'.
+      let sameGC c = show (UChar.generalCategory c) == show (Char.generalCategory c)
+      traverse_ (`shouldSatisfy` sameGC) [minBound..maxBound]
+  describe' "Character classification" do
+    it "isAlpha" do
+      let same c = UChar.isAlpha c == Char.isAlpha c
+      traverse_ (`shouldSatisfy` same) [minBound..maxBound]
```
Comment (harendra-kumar) [inline test/Unicode/CharSpec.hs:43]: This pattern is repeating in all tests, can be factored out as a function.

File: unicode-data.cabal [hunk 3]
```diff
@@ -110,10 +112,43 @@ executable ucd2haskell
...
+  if impl(ghc >= 9.2.1)
+    cpp-options: -DCOMPATIBLE_GHC_UNICODE
+  default-language: Haskell2010
+
+benchmark Unicode.Char
+  import: default-extensions, compile-options
+  type: exitcode-stdio-1.0
+  hs-source-dirs: bench
+  main-is: Main.hs
+  build-depends:
```
Comment (harendra-kumar) [inline unicode-data.cabal:148]: Remove additional whitespace at the end of the line.

File: bench/Main.hs [hunk 4]
```diff
@@ -0,0 +1,154 @@
+import Data.Ix (Ix(..))
+import Control.DeepSeq (NFData, deepseq)
+import Test.Tasty.Bench
```
Comment (harendra-kumar) [inline bench/Main.hs:3]: Sort the imports, use explicit imports.

```
import Control.DeepSeq (NFData, deepseq)
import Data.Ix (Ix(..))
import Test.Tasty.Bench (bgroup, bench, nf, defaultMain)
```

**Review by harendra-kumar — CHANGES_REQUESTED** [review]
Overall looks pretty good. I have some minor comments.

We can add a `cabal docspec` CI to check the properties in haddock. We have a similar CI in the streamly repo.

File: bench/Main.hs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+import qualified Unicode.Char.General as G
+import qualified Unicode.Char.Identifiers as I
+import qualified Unicode.Char.Normalization as N
+
+main :: IO ()
+main = defaultMain
+  [ bgroup "Unicode.Char.Case"
+    [ bgroup "isLower"
+      [ bench "base"         $ nf (fold_ B.isLower) (minBound, maxBound)
+      , bench "unicode-data" $ nf (fold_ C.isLower) (minBound, maxBound)
```
[the remark this answers is not in the document]
Comment (Bodigrim) [reply bench/Main.hs:16]: I'd suggest to use [`bcompare`](https://hackage.haskell.org/package/tasty-bench-0.3.1/docs/Test-Tasty-Bench.html#v:bcompare) to report relative speed up immediately. Here is a real-world example: https://github.com/haskell-perf/sequences/pull/28/files#diff-c8a34a56bc42d89667f2f638e224462c48e2a471668e144b1b621d7f448afdd8R187

File: lib/Unicode/Char/General.hs [hunk 6]
```diff
@@ -75,29 +91,391 @@ where
...
+    * @Lowercase@ [property](https://www.unicode.org/reports/tr44/#Lowercase)
+    * 'TitlecaseLetter' (@Lt@)
+    * 'ModifierLetter' (@Lm@)
+    * 'OtherLetter' (@Lo@)
+    * 'LetterNumber' (@Nl@)
+    * @Other_Alphabetic@ [property](https://www.unicode.org/reports/tr44/#Other_Alphabetic)
+-}
+{-# INLINE isAlphabetic #-}
+isAlphabetic :: Char -> Bool
+isAlphabetic = P.isAlphabetic
```
[the remark this answers is not in the document]
Comment (Bodigrim) [reply lib/Unicode/Char/General.hs:256]: Fine with me, no strong opinion.

File: Changelog.md [hunk 7]
```diff
@@ -1,5 +1,23 @@
...
+## Next
+
+- Parser:
+    - Fix parsing of ranges in `UnicodeData.txt`.
+    - Generate `lib/Unicode/Internal/Char/UnicodeData/GeneralCategory.hs`.
+- Library:
+    - Support for big-endian architectures.
+    - Add `GeneralCategory` data type and corresponding `generalCategoryAbbr, generalCategory` functions.
+    - Add the following functions: `isAlpha, isAlphaNum, isControl, isMark, isNumber, isPrint, isPunctuation, isSeparator, isSymbol`.
+    - **Breaking change:** `isLetter` is renamed `isAlphabetic` and `isSpace` is renamed `isWhiteSpace`.
```
Comment (adithyaov) [inline Changelog.md:12]: - Added `isAlphabetic` and `isWhiteSpace`.
- Changed the behavior of `isLetter` and `isSpace` to match base's `Data.Char` behavior.

File: bench/Main.hs [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    -- Character classification
+    [ bgroup "generalCategory"
+      [ bench "base"         $ nf (fold_ (show . B.generalCategory)) (minBound, maxBound)
+      , bench "unicode-data" $ nf (fold_ (show . G.generalCategory)) (minBound, maxBound)
+      ]
+    , bgroup "isAlpa"
+      [ bench "base"         $ nf (fold_ B.isAlpha) (minBound, maxBound)
+      , bench "unicode-data" $ nf (fold_ G.isAlpha) (minBound, maxBound)
+      ]
+    , bgroup "isAlpabetic"
```
Comment (adithyaov) [inline bench/Main.hs:33]: typo.

File: exe/Parser/Text.hs [hunk 9]
```diff
@@ -144,6 +149,41 @@ bitMapToAddrLiteral = map (chr . toByte . padTo8) . unfoldr go
...
+
+    go :: a -> Char
+    go = chr . fromIntegral . toWord8
+
+    toWord8 :: a -> Word8
+    toWord8 a = let w = fromEnum a in if 0 <= w && w <= 0xff
+        then fromIntegral w
+        else error $ "Cannot convert to Word8: " <> show a
+
+    padWord = (<> "\0\0\0\0\0\0\0\0")
```
Comment (adithyaov) [inline exe/Parser/Text.hs:186]: Why `padWord`?

File: lib/Unicode/Char/General.hs [hunk 10]
```diff
@@ -9,9 +9,25 @@
 --
 module Unicode.Char.General
     (
-    -- * Character Properties
-      isLetter
+    -- * Unicode general categories
+      GeneralCategory(..)
+    , generalCategoryAbbr
```
Comment (adithyaov) [inline lib/Unicode/Char/General.hs:14]: Do these follow the module classification according to tr44?

**Comment by adithyaov** [comment]
@wismill I apologize for the delayed review. The PR looks good overall. I don't have any major comments other than the minor scrutiny already mentioned. I've added in one small question along as well.

You might need to rebase and resolve conflicts.

I can send a PR addressing this minor scrutiny to this branch in your repo.

File: exe/Parser/Text.hs (hunk 9 above)
Comment (wismill) [reply exe/Parser/Text.hs:186]: Remain from previous subword processing attempt.

File: Changelog.md [hunk 11]
```diff
@@ -1,5 +1,23 @@
...
+    - **Breaking change:** `isLetter` and `isSpace` now match the base’s Data.Char behaviour.
+    - Add `lookupIntN` in `Unicode.Internal.Bits` for low-level stuff.
+    - Re-export some functions from `Data.Char` in order to make `Unicode.Char` a drop-in replacement.
+- Test:
+   - Add tests for `Unicode.Char` to ensure we get the same result than `Data.Char`.
+     Note that this true only if the compiler used have have the same Unicode version as this package.
+- Add Benchmark.
+
```
[the remark this answers is not in the document]
Comment (wismill) [reply Changelog.md:16]: Done

File: Changelog.md [hunk 12]
```diff
@@ -1,5 +1,23 @@
...
+- Parser:
+    - Fix parsing of ranges in `UnicodeData.txt`.
+    - Generate `lib/Unicode/Internal/Char/UnicodeData/GeneralCategory.hs`.
+- Library:
+    - Support for big-endian architectures.
+    - Add `GeneralCategory` data type and corresponding `generalCategoryAbbr, generalCategory` functions.
+    - Add the following functions: `isAlpha, isAlphaNum, isControl, isMark, isNumber, isPrint, isPunctuation, isSeparator, isSymbol`.
+    - **Breaking change:** `isLetter` is renamed `isAlphabetic` and `isSpace` is renamed `isWhiteSpace`.
+    - **Breaking change:** `isLetter` and `isSpace` now match the base’s Data.Char behaviour.
+    - Add `lookupIntN` in `Unicode.Internal.Bits` for low-level stuff.
```
[the remark this answers is not in the document]
Comment (wismill) [reply Changelog.md]: Done

File: bench/Main.hs [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+import qualified Data.Char as B
+import qualified Unicode.Char.Case as C
+import qualified Unicode.Char.General as G
+import qualified Unicode.Char.Identifiers as I
+import qualified Unicode.Char.Normalization as N
+
+main :: IO ()
+main = defaultMain
+  [ bgroup "Unicode.Char.Case"
+    [ bgroup "isLower"
```
[the remark this answers is not in the document]
Comment (wismill) [reply bench/Main.hs:14]: Sorry, I do not have time to fix this. Anyway I think using a formatter (i.e. formulu) is a good idea to enforce coding style. Would you choose one for this project?

File: bench/Main.hs [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+      , bench "unicode-data" $ nf (fold_ C.isUpper) (minBound, maxBound)
+      ]
+    ]
+  , bgroup "Unicode.Char.General"
+    -- Character classification
+    [ bgroup "generalCategory"
+      [ bench "base"         $ nf (fold_ (show . B.generalCategory)) (minBound, maxBound)
+      , bench "unicode-data" $ nf (fold_ (show . G.generalCategory)) (minBound, maxBound)
+      ]
+    , bgroup "isAlpa"
```
[the remark this answers is not in the document]
Comment (wismill) [reply bench/Main.hs]: Fixed

File: bench/Main.hs [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+      [ bench "base"         $ nf (fold_ B.isAlpha) (minBound, maxBound)
+      , bench "unicode-data" $ nf (fold_ G.isAlpha) (minBound, maxBound)
+      ]
+    , bgroup "isAlpabetic"
+      [ bench "unicode-data" $ nf (fold_ G.isAlphabetic) (minBound, maxBound)
+      ]
+    , bgroup "isAlpaNum"
```
[the remark this answers is not in the document]
Comment (wismill) [reply bench/Main.hs]: Fixed

File: lib/Unicode/Char/General.hs [hunk 16]
```diff
@@ -75,29 +91,391 @@ where
...
+generalCategoryAbbr ModifierSymbol       = "Sk"
+generalCategoryAbbr OtherSymbol          = "So"
+generalCategoryAbbr Space                = "Zs"
+generalCategoryAbbr LineSeparator        = "Zl"
+generalCategoryAbbr ParagraphSeparator   = "Zp"
+generalCategoryAbbr Control              = "Cc"
+generalCategoryAbbr Format               = "Cf"
+generalCategoryAbbr Surrogate            = "Cs"
+generalCategoryAbbr PrivateUse           = "Co"
+generalCategoryAbbr NotAssigned          = "Cn"
```
[the remark this answers is not in the document]
Comment (wismill) [reply lib/Unicode/Char/General.hs]: Fixed

File: test/Unicode/CharSpec.hs (hunk 2 above)
Comment (wismill) [reply test/Unicode/CharSpec.hs]: Fixed

File: unicode-data.cabal (hunk 3 above)
Comment (wismill) [reply unicode-data.cabal]: Fixed

File: bench/Main.hs (hunk 4 above)
Comment (wismill) [reply bench/Main.hs]: Fixed

File: Changelog.md (hunk 7 above)
Comment (wismill) [reply Changelog.md]: Fixed

File: bench/Main.hs (hunk 8 above)
Comment (wismill) [reply bench/Main.hs]: Fixed

File: bench/Main.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (wismill) [reply bench/Main.hs]: Fixed repetition

File: lib/Unicode/Char.hs [hunk 17]
```diff
@@ -30,9 +30,27 @@ module Unicode.Char
     , module Unicode.Char.Case
     , module Unicode.Char.Normalization
     , module Unicode.Char.Identifiers
+
+    -- * Re-export
+    -- ** Numbers
+    , isDigit
+    , isOctDigit
+    , isHexDigit
```
[the remark this answers is not in the document]
Comment (wismill) [reply lib/Unicode/Char.hs]: Fixed

File: lib/Unicode/Char.hs [hunk 18]
```diff
@@ -30,9 +30,27 @@ module Unicode.Char
...
+    -- ** Subranges
+    , isAscii
+    , isLatin1
+    , isAsciiUpper
+    , isAsciiLower
```
[the remark this answers is not in the document]
Comment (wismill) [reply lib/Unicode/Char.hs]: Fixed

File: lib/Unicode/Char/General.hs [hunk 19]
```diff
@@ -75,29 +91,391 @@ where
...
+Alias of 'isLetter'.
+
+__Note:__ this function is /not/ equivalent to 'isAlphabetic'.
+See the description of 'isAlphabetic' for further details.
+
+prop> isAlpha c == Data.Char.isAlpha c
+-}
+{-# INLINE isAlpha #-}
+isAlpha :: Char -> Bool
+isAlpha = isLetter
```
[the remark this answers is not in the document]
Comment (wismill) [reply lib/Unicode/Char/General.hs:218]: Fixed

File: bench/Main.hs (hunk 13 above)
[the remark this answers is not in the document]
Comment (harendra-kumar) [reply bench/Main.hs:15]: `fourmolu` with appropriate config might work, but someone will have to do that work as well. We can leave it as it is for now.

**Comment by harendra-kumar** [comment]
The CI's are failing with this message:

> The following files are committed to the git repository but do not exist in the source distribution.
> > .editorconfig
> Please consider adding them to your cabal file under 'extra-source-files' or 'extra-doc-files'.
> If you do not want to add them in the source distribution then add them to .packcheck.ignore file at the root of the git repository.

You can add .editorconfig to .packcheck.ignore since we do not want to pack it in the distribution.

**Comment by harendra-kumar** [comment]
@adithyaov can you suggest a fix for appveyor CI failure. I think you fixed a similar issue elsewhere.

**Comment by harendra-kumar** [comment]
@wismill this looks good for merge. You can squash any commits if required before the merge.

File: bench/Main.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (wismill) [reply bench/Main.hs]: Added bcompare

**Comment by wismill** [comment]
New benchmark with relative speedup:

```
All
  Unicode.Char.Case
    isLower
      base:           OK (0.17s)
         21 ms ± 1.4 ms
      unicode-data:   OK (0.12s)
        3.9 ms ± 340 μs, 0.18x
    isUpper
      base:           OK (0.15s)
         22 ms ± 1.7 ms
      unicode-data:   OK (0.23s)
        3.7 ms ± 168 μs, 0.17x
  Unicode.Char.General
    generalCategory
      base:           OK (0.37s)
        124 ms ± 4.2 ms
      unicode-data:   OK (0.32s)
        106 ms ± 5.5 ms, 0.85x
    isAlpha
      base:           OK (0.15s)
         21 ms ± 1.4 ms
      unicode-data:   OK (0.25s)
        3.9 ms ± 180 μs, 0.18x
    isAlphabetic
      unicode-data:   OK (0.16s)
        312 μs ±  23 μs
    isAlphaNum
      base:           OK (0.15s)
         22 ms ± 1.5 ms
      unicode-data:   OK (0.14s)
        4.6 ms ± 337 μs, 0.21x
    isControl
      base:           OK (0.15s)
         22 ms ± 1.5 ms
      unicode-data:   OK (0.55s)
        4.5 ms ± 365 μs, 0.21x
    isLetter
      base:           OK (0.32s)
         21 ms ± 904 μs
      unicode-data:   OK (0.12s)
        3.9 ms ± 342 μs, 0.18x
    isMark
      base:           OK (0.16s)
         23 ms ± 2.2 ms
      unicode-data:   OK (0.27s)
        4.3 ms ± 175 μs, 0.19x
    isPrint
      base:           OK (0.34s)
         22 ms ± 1.4 ms
      unicode-data:   OK (0.13s)
        4.2 ms ± 378 μs, 0.18x
    isPunctuation
      base:           OK (0.15s)
         21 ms ± 2.1 ms
      unicode-data:   OK (0.13s)
        4.3 ms ± 351 μs, 0.20x
    isSeparator
      base:           OK (0.34s)
         23 ms ± 847 μs
      unicode-data:   OK (0.12s)
        4.0 ms ± 369 μs, 0.17x
    isSpace
      base:           OK (2.97s)
         11 ms ±  60 μs
      unicode-data:   OK (0.15s)
        4.8 ms ± 381 μs, 0.42x
    isSymbol
      base:           OK (0.33s)
         22 ms ± 1.0 ms
      unicode-data:   OK (0.14s)
        4.4 ms ± 377 μs, 0.20x
    isWhiteSpace
      unicode-data:   OK (0.16s)
        312 μs ±  22 μs
    isHangul
      unicode-data:   OK (0.16s)
        312 μs ±  21 μs
    isHangulLV
      unicode-data:   OK (0.16s)
        312 μs ±  22 μs
    isJamo
      unicode-data:   OK (0.16s)
        312 μs ±  22 μs
    jamoLIndex
      unicode-data:   OK (0.16s)
        312 μs ±  22 μs
    jamoVIndex
      unicode-data:   OK (0.16s)
        312 μs ±  21 μs
    jamoTIndex
      unicode-data:   OK (0.16s)
        312 μs ±  21 μs
  Unicode.Char.Identifiers
    isIDContinue
      unicode-data:   OK (0.16s)
        312 μs ±  23 μs
    isIDStart
      unicode-data:   OK (0.16s)
        312 μs ±  21 μs
    isXIDContinue
      unicode-data:   OK (0.16s)
        312 μs ±  21 μs
    isXIDStart
      unicode-data:   OK (0.16s)
        312 μs ±  21 μs
    isPatternSyntax
      unicode-data:   OK (0.16s)
        312 μs ±  22 μs
    isPatternWhitespace
      unicode-data:   OK (0.16s)
        312 μs ±  21 μs
  Unicode.Char.Normalization
    isCombining
      unicode-data:   OK (0.16s)
        312 μs ±  21 μs
    combiningClass
      unicode-data:   OK (0.19s)
        2.9 ms ± 221 μs
    isCombiningStarter
      unicode-data:   OK (0.16s)
        312 μs ±  21 μs
    isDecomposable
      Canonical
        unicode-data: OK (0.16s)
          624 μs ±  42 μs
      Kompat
        unicode-data: OK (0.16s)
          625 μs ±  46 μs
    decomposeHangul
      unicode-data:   OK (0.16s)
        623 μs ±  43 μs
  Unicode.Char.Numeric
    isNumber
      base:           OK (0.16s)
         23 ms ± 1.7 ms
      unicode-data:   OK (0.15s)
        4.8 ms ± 345 μs, 0.21x

All 48 tests passed (12.06s)
```

File: lib/Unicode/Char/General.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (wismill) [reply lib/Unicode/Char/General.hs:246]: My reasoning on this is:
- Predicates are named from their property in Unicode. They may have no equivalent in `Data.Char`.
- `base` predicates are kept unchanged.
- Having `Unicode.Char` a drop-in replacement of `Data.Char` make it possible to ensure a specific version of Unicode rather than depend on GHC’s Unicode version.

I agree that having predicates that differ only for a few characters is confusing. I would rather use the ones from Unicode when possible. Therefore I propose the following:
- Move `isAlpha`, `isLetter` and `isSpace` to a new module `Unicode.Char.General.Compat`.
- Re-export `Unicode.Char.General.Compat` in `Unicode.Char`.

File: lib/Unicode/Char/General.hs [hunk 20]
```diff
@@ -75,29 +91,391 @@ where
...
+{- | Returns 'True' for any whitespace characters, and the control
+characters @\\t@, @\\n@, @\\r@, @\\f@, @\\v@.
+
+See: [White_Space](https://www.unicode.org/reports/tr44/#White_Space).
+
+__Note:__ 'isWhiteSpace' is /not/ equivalent to 'isSpace'.
+-}
+{-# INLINE isWhiteSpace #-}
+isWhiteSpace :: Char -> Bool
+isWhiteSpace = P.isWhite_Space
```
[the remark this answers is not in the document]
Comment (wismill) [reply lib/Unicode/Char/General.hs:424]: See my comment about `isAlphabetic`.

File: lib/Unicode/Char/General.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (harendra-kumar) [reply lib/Unicode/Char/General.hs:246]: Sounds good. I am fine with it even the way it is now, you can decide whichever way you like. We can mark them `Compat` APIs in the haddock and suggest the better/recommended alternative. Anyway, the benchmarks look favorable for the new APIs, so that's another reason to use those.

**Comment by harendra-kumar** [comment]
The benchmarks look pretty good compared to `base`. It may be a good idea for GHC/base to use the Haskell native `unicode-data` instead of using FFI to C library functions. And we have a good chunk of unicode functionality supported now, I am sure we will keep adding more stuff and one day this will be comprehensive.

Good stuff @wismill !

File: lib/Unicode/Char/General.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (wismill) [reply lib/Unicode/Char/General.hs:239]: I moved the functions to a compatiblity module.

**Comment by wismill** [comment]
Ok, I think this PR is almost ready. Should I squash all the commits? Should I bump the package’s version to 0.3?

**Comment by harendra-kumar** [comment]
You can either squash them into a single commit or maybe squash them into a few logically related commits, or just squash the trivial/fixup type commits and leave the rest as it is, whatever you deem fit. 

We can bump the package version, but let's give it a soak time of a week before we upload to hackage. We can also generate a bench-show report with a side-by-side benchmark comparison with base and put it in the readme, like we have in unicode-transforms.

**Comment by adithyaov** [comment]
This needs to be added to the stack config.
```
flags:
  mintty:
    Win32-2-13-1: false
```
See https://stackoverflow.com/questions/70045586/could-not-find-module-system-console-mintty-win32-when-compiling-test-framework

**Comment by wismill** [comment]
> We can also generate a bench-show report with a side-by-side benchmark comparison with base and put it in the readme, like we have in unicode-transforms.
Unfortunately it seems that `bench-show` does not support `tasty-bench` output. I will add an excerpt of the current output.

**Comment by adithyaov** [comment]
> Unfortunately it seems that bench-show does not support tasty-bench output. I will add an excerpt of the current output.

I'll need to update `bench-show` and generate the benchmarks. I'll need to do this as a release task anyway.

**Comment by wismill** [comment]
> > Unfortunately it seems that bench-show does not support tasty-bench output. I will add an excerpt of the current output.
> 
> I'll need to update `bench-show` and generate the benchmarks. I'll need to do this as a release task anyway.

Ok. Apart from this and GHC 7.10.3 failing, I think this PR is ready to merge.

File: lib/Unicode/Char/General.hs [hunk 21]
```diff
@@ -75,29 +91,391 @@ where
...
+
+{-| Unicode General Categories.
+
+These classes are defined in the
+[Unicode Character Database](http://www.unicode.org/reports/tr44/tr44-14.html#GC_Values_Table),
+part of the Unicode standard
+
+__Note:__ the classes must be in the same order they are listed in the Unicode Standard,
+because some functions (e.g. 'generalCategory') rely on the 'Enum' instance.
+-}
```
[the remark this answers is not in the document]
Comment (wismill) [reply lib/Unicode/Char/General.hs:114]: To be solved by formatter.

**Comment by wismill** [comment]
Let’s merge and continue work in specific issues.
